### ISSUE TITLE
docker workflow

### ISSUE THREAD

**Issue by dghewrhrwehw** [mac-s-g/react-json-view#135]
hey.. love your work and curious how you delineate your usage of shell scripts (`./docker/`, `./entrypoints/`) and npm (e.g. `"npm run dev-hot"`). I'm interested in using some ideas from this repo to organize my own work. Are there pros/cons to running in the docker container?
Always trying to improve my workflow!

(p.s. happy to repost this on stackoverflow or the like)

**Comment by mac-s-g** [comment]
hey, i'm really glad you reached out!

There are some awesome benefits that come with adding docker to your
workflow.

I don't have time to give you a breakdown tonight, but I can follow up in
the morning.

Maybe I can enumerate some of that conversation into a readme and post it
in `/docker/` to help share with others.

Mac

**Comment by mac-s-g** [comment]
I'm gonna attempt to motivate this discussion with some high level info and then drill in a bit.

#### What is Docker
docker manages software "containers".
a container is a stack of "images" that define an environment from the OS up.
for example, the container used with this project has the following composition:

```
[ RJV node modules ]
[ node 8.4.0       ]
[ linux debian     ]
```

#### Motivations for Docker
 * _environmental consistency_
   * this means I can completely replicate an environment for any build, on any machine
 * _portability_
   * it's easy to define an environment and share it with other developers.  
   * I don't need to list steps for installing software - I can just write and share a dockerfile
 * _module fatigue_
   * I can easily add/update/remove frameworks or node modules without worrying about messing up my file system.  All packages are confined within a container, which I can easily update, rebuild or remove.

#### RJV Docker Workflow

the `Dockerfile` defines exactly how a container should be composed for an application.  Any docker-integrated workflow starts here.

Here is the [RJV Dockerfile](https://github.com/mac-s-g/react-json-view/blob/master/Dockerfile)

Notice [here](https://github.com/mac-s-g/react-json-view/blob/master/Dockerfile#L1) that my container is built on top of a node container.  If you track down that node container on docker-hub, you'll see that it's built on top of a debian container.  Also notice that `node_modules` are installed directly into my container [here](https://github.com/mac-s-g/react-json-view/blob/master/Dockerfile#L6-L7).  

When I run `./docker/build-container.sh`, docker produces a container including all the `node_modules` listed in my `package.json` file.

When I run `./docker/dev-server.sh`, a few things happen:
 1. docker [runs my container and mounts some source files](https://github.com/mac-s-g/react-json-view/blob/master/docker/dev-server.sh#L16-L20) into the container 
    * "mounting" is like a soft copy.  my files are linked inside the container's file system so local edits propagate to the container.
 2. docker [exposes a port](https://github.com/mac-s-g/react-json-view/blob/master/docker/dev-server.sh#L23) on the container
    * this allows me to configure my app to listen and respond to to traffic at http://localhost:2000
    * the port is arbitrary
 3. An [entrypoint will be invoked](https://github.com/mac-s-g/react-json-view/blob/master/docker/dev-server.sh#L24) once the container is running
    * With my container up and running, the [entrypoint simply runs webpack dev server](https://github.com/mac-s-g/react-json-view/blob/master/entrypoints/dev-server.sh#L6).

If you understand those steps, then you'll understand any of the docker scripts in `/docker/`.  They do the same thing as the dev-server script but call different entrypoints.

here is a workflow comparison with and without  docker:

Task|Without Docker|With Docker
|:---|:---|:---
install node modules|update package.json<br/>`npm install`|update package.json<br/>`./docker/build-container.sh`
run dev server|`npm run dev:hot`|`./docker/dev-server.sh`
run build|`npm run build`|`./docker/build.sh`

#### is it worth it?

`/docker/` and `/entrypoint/` scripts come with some overhead.  Is it really worth the work?

Remember the problems that docker is solving.  Without docker, I have to be much more skeptical when testing contributions from the community.  

What if a contributor is using a different version of node or npm?  
What if a contributor has a global package installed that's affecting the behavior of their code?  

running the app inside a container ensures that environmental inconsistencies like those above will not lead to bugs in production.

#### more examples

[github-help-wanted](https://github.com/mac-s-g/github-help-wanted) is a newer project that has a simpler docker workflow.

@dghewrhrwehw  does this help?
What other questions do you have?
Are you convinced or skeptical about the advantages of docker?

**Comment by mac-s-g** [comment]
I migrated this documentation to [./docker/README.md](https://github.com/mac-s-g/react-json-view/blob/master/docker/README.md) with pr #146 and #147.

i'm closing for now, but i'm happy to continue the discussion!

**Comment by buchanae** [comment]
Docker, a containerized filesystem, webpack, npm, entrypoints, babel, transpiled CSS, source maps, a dozen bash scripts, nearly a dozen npm commands. All for a single React component. Gotta say, I miss the days where you just opened an HTML file. 

Anyway, I came here to contribute and ran into 30-60 minutes of trouble getting a dev server running. I think your dev server config now requires the Docker container (because of the "/react" basepath). Unfortunately, I still don't have a working server, so I'm giving up :(

**Comment by mac-s-g** [comment]
@buchanae - I hear you!

thanks for pointing out the path related dependencies on docker.  I resolved those with #148.

I also tested the standard workflow, and, well... it's broken.  I fixed it with #149.  
are you interested in pulling the latest code and trying one more time?
```
npm install --save-dev  #I had to update some dependencies 
npm run dev:hot
``` 

It's working for me now, so I hope your problems will also be solved.  I also hope the docker tools haven't discouraged many contributors.  I'd like the tools to be available but not invasive.  I believe in the benefits I've seen from docker, but it's not worth discouraging contributions from the community.
